x/farm/keeper/farm.go

```diff
@@ -0,0 +1,244 @@
+package keeper
+
+import (
+	sdk "github.com/cosmos/cosmos-sdk/types"
+	sdkerrors "github.com/cosmos/cosmos-sdk/types/errors"
+
+	"github.com/cosmosquad-labs/squad/v3/x/farm/types"
+)
+
+func (k Keeper) Farm(ctx sdk.Context, farmerAddr sdk.AccAddress, coin sdk.Coin) (withdrawnRewards sdk.Coins, err error) {
+	farmingReserveAddr := types.DeriveFarmingReserveAddress(coin.Denom)
+	if err := k.bankKeeper.SendCoins(
+		ctx, farmerAddr, farmingReserveAddr, sdk.NewCoins(coin)); err != nil {
+		return nil, err
+	}
+
+	farm, found := k.GetFarm(ctx, coin.Denom)
+	if !found {
+		farm = k.initializeFarm(ctx, coin.Denom)
+	}
+
+	position, found := k.GetPosition(ctx, farmerAddr, coin.Denom)
+	if !found {
+		k.incrementFarmPeriod(ctx, coin.Denom)
+		position = types.Position{
+			Farmer:        farmerAddr.String(),
+			Denom:         coin.Denom,
+			FarmingAmount: sdk.ZeroInt(),
+		}
+	} else {
+		withdrawnRewards, err = k.withdrawRewards(ctx, position)
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	farm, _ = k.GetFarm(ctx, coin.Denom)
+	farm.TotalFarmingAmount = farm.TotalFarmingAmount.Add(coin.Amount)
+	k.SetFarm(ctx, coin.Denom, farm)
+
+	position.FarmingAmount = position.FarmingAmount.Add(coin.Amount)
+	k.updatePosition(ctx, position)
+
+	// TODO: emit an event
+
+	return withdrawnRewards, nil
+}
+
+func (k Keeper) Unfarm(ctx sdk.Context, farmerAddr sdk.AccAddress, coin sdk.Coin) (withdrawnRewards sdk.Coins, err error) {
+	position, found := k.GetPosition(ctx, farmerAddr, coin.Denom)
+	if !found {
+		return nil, sdkerrors.Wrap(sdkerrors.ErrNotFound, "position not found")
+	}
+	if position.FarmingAmount.LT(coin.Amount) {
+		return nil, sdkerrors.Wrapf(sdkerrors.ErrInsufficientFunds, "not enough farming amount")
+	}
+
+	withdrawnRewards, err = k.withdrawRewards(ctx, position)
+	if err != nil {
+		return nil, err
+	}
+
+	position.FarmingAmount = position.FarmingAmount.Sub(coin.Amount)
+	if position.FarmingAmount.IsZero() {
+		k.DeletePosition(ctx, farmerAddr, coin.Denom)
+	} else {
+		k.updatePosition(ctx, position)
+	}
+
+	farm, found := k.GetFarm(ctx, coin.Denom)
+	if !found {
+		return nil, sdkerrors.Wrap(sdkerrors.ErrNotFound, "farm not found")
+	}
+	farm.TotalFarmingAmount = farm.TotalFarmingAmount.Sub(coin.Amount)
+	k.SetFarm(ctx, coin.Denom, farm)
+
+	farmingReserveAddr := types.DeriveFarmingReserveAddress(coin.Denom)
+	if err := k.bankKeeper.SendCoins(ctx, farmingReserveAddr, farmerAddr, sdk.NewCoins(coin)); err != nil {
+		return nil, err
+	}
+
+	// TODO: emit an event
+
+	return withdrawnRewards, nil
+}
+
+func (k Keeper) Harvest(ctx sdk.Context, farmerAddr sdk.AccAddress, denom string) (withdrawnRewards sdk.Coins, err error) {
+	position, found := k.GetPosition(ctx, farmerAddr, denom)
+	if !found {
+		return nil, sdkerrors.Wrap(sdkerrors.ErrNotFound, "position not found")
+	}
+
+	// TODO: prevent incrementing too many periods?
```

Comment:
Is this when a farmer calls `Harvest` too many times?

Reply:
Yes, right. Now that I think about it, I don't think this will cause a big problem at all, since incrementing too many periods doesn't much affect the system.